src/backend/gl/src/lib.rs

```diff
@@ -315,6 +316,162 @@ impl PhysicalDevice {
+        if private_caps.map {
+            buffer_role_memory_types.push(hal::MemoryType {
+                properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED | memory::Properties::COHERENT,
+                heap_index: CPU_VISIBLE_HEAP,
+            });
+            buffer_role_memory_types.push(hal::MemoryType {
+                properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED,
+                heap_index: CPU_VISIBLE_HEAP,
+            });
+        } else if private_caps.emulate_map {
+            buffer_role_memory_types.push(hal::MemoryType {
+                properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED,
+                heap_index: CPU_VISIBLE_HEAP,
+            });
+        }
+        buffer_role_memory_types.push(hal::MemoryType {
+            properties: memory::Properties::DEVICE_LOCAL,
+            heap_index: DEVICE_LOCAL_HEAP,
+        });
+
+        let mut memory_types = Vec::new();
+
+        for &buffer_role_memory_type in &buffer_role_memory_types {
+            if private_caps.buffer_role_change {
+                if info.is_webgl() {
+                    // For security reasons, WebGL does not allow "element array buffers" to be used
+                    // as any other kind of buffer (even when `buffer_role_change` is true).  We
+                    // need to provide unique types of memory specifically for buffers with INDEX
+                    // usage if we are on WebGL.
+                    memory_types.push((
+                        buffer_role_memory_type,
+                        MemoryRole::Buffer {
+                            allowed_usage: buffer::Usage::INDEX,
+                            target: glow::ELEMENT_ARRAY_BUFFER,
+                        }
+                    ));
+                    memory_types.push((
+                        buffer_role_memory_type,
+                        MemoryRole::Buffer {
+                            allowed_usage: buffer::Usage::TRANSFER_SRC
+                                | buffer::Usage::TRANSFER_DST
+                                | buffer::Usage::UNIFORM_TEXEL
+                                | buffer::Usage::STORAGE_TEXEL
+                                | buffer::Usage::UNIFORM
+                                | buffer::Usage::STORAGE
+                                | buffer::Usage::VERTEX
+                                | buffer::Usage::INDIRECT,
+                            target: glow::PIXEL_PACK_BUFFER,
+                        }
+                    ));
+                } else {
+                    // If we have `buffer_role_change` capability and are not on WebGL, a buffer can
+                    // be used for any role.
+                    memory_types.push((
+                        buffer_role_memory_type,
+                        MemoryRole::Buffer {
+                            allowed_usage: buffer::Usage::TRANSFER_SRC
```

Comment:
could we use `buffer::Usage::all()` instead?

Reply:
Oh sure thing I just forgot it existed :P